packages/flutter/lib/src/material/text_field.dart

```diff
@@ -1193,7 +1232,26 @@ class _TextFieldState extends State<TextField> with RestorationMixin implements
       setState(() {
         _isHovering = hovering;
       });
+      _statesController.update(MaterialState.hovered, _isHovering);
+    }
+  }
+
+  // Material states controller.
+  MaterialStatesController? _internalStatesController;
```

Comment:
That is a good point we are currently not using it internally. We do keep track of our `_materialState` using a private member https://github.com/flutter/flutter/blob/ff5b5b5fa6f35b717667719ddfdb1521d8bdd05a/packages/flutter/lib/src/material/text_field.dart#L1215-L1222 maybe we should just use the controller to track this in one place https://github.com/flutter/flutter/pull/133977/commits/c1d4cdf76e5a6507de45f41aa0ee2691511178c9.

Reply:
The purpose of this change is to let developers monitor material states changes, but changing the material states from outside of the `TextField` class shouldn't be allowed, since the source of the truth is `TextFieldState` I think?

Reply:
It does makes sense that TextField would be the source of truth for its material state, but I also see the use-case for being able to programmatically update a TextField's material state to add support for more states.

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MaterialApp(home: Home()));
}

class PressedInputField extends StatefulWidget {
  const PressedInputField({
    super.key,
    this.style,
    required this.pressed,
    required this.onPressed,
    required this.onReleased,
  });

  final bool pressed;
  final TextStyle? style;
  final VoidCallback? onPressed;
  final VoidCallback? onReleased;

  @override
  State<PressedInputField> createState() => _PressedInputFieldState();
}

class _PressedInputFieldState extends State<PressedInputField> {
  late final MaterialStatesController statesController;
  final TextEditingController controller = TextEditingController(text: 'some text');

  @override
  void initState() {
    super.initState();
    statesController = MaterialStatesController(
        <MaterialState>{if (widget.pressed) MaterialState.pressed});
  }

  @override
  void dispose() {
    statesController.dispose();
    controller.dispose();
    super.dispose();
  }

  @override
  void didUpdateWidget(PressedInputField oldWidget) {
    super.didUpdateWidget(oldWidget);
    if (widget.pressed != oldWidget.pressed) {
      statesController.update(MaterialState.pressed, widget.pressed);
    }
  }

  @override
  Widget build(BuildContext context) {
    return Listener(
      behavior: HitTestBehavior.translucent,
      onPointerDown: (PointerDownEvent event) {
        widget.onPressed?.call();
      },
      child: TextField(
        controller: controller,
        statesController: statesController,
        style: widget.style,
        onTap: () {
          widget.onReleased?.call(); 
        },
      ),
    );
  }
}

class Home extends StatefulWidget {
  const Home({super.key});

  @override
  State<Home> createState() => _HomeState();
}

class _HomeState extends State<Home> {
  bool pressed = false;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: PressedInputField(
          pressed: pressed,
          style: MaterialStateTextStyle.resolveWith((Set<MaterialState> states) {
            if (states.contains(MaterialState.pressed)) {
              return const TextStyle(color: Colors.white, backgroundColor: Colors.indigo);
            }
            return const TextStyle(color: Colors.teal);
          }),
          onPressed: () {
            setState(() {
              pressed = !pressed;
            });
          },
          onReleased: () {
            setState(() {
              pressed = !pressed;
            });
          }
        ),
      ),
    );
  }
}
```

I'm okay with holding off on that functionality until it is requested, but I also think when a user updates state through the controller using `_myController.update(MaterialState.pressed, true)` they might expect that the receiving widget reflects the value change, similar to other controllers like `TextEditingController` and doing `_myController.value = myNewValue`.

Reply:
> when a user updates state through the controller using _myController.update(MaterialState.pressed, true) they might expect that the receiving widget reflects the value change, similar to other controllers like TextEditingController and doing _myController.value = myNewValue.

Do you mean if a `TextField` has focus but if the user calls `update(.focused, false)` on the controller then the text field should tell its focus node to unfocus? That sounds pretty strange to me.

Reply:
I think when the user calls `update(.focused, false)` they exepect `TextField` internal `Set<MaterialState>` to reflect the update. This way a user can add support for more states if they wanted to, for example passing a `MaterialStateTextStyle` that supports `MaterialState.pressed`. I agree that it would be strange for `TextField` to update its actual state (FocusNode) to be unfocused on changes to `MaterialStateController`.

I'm not entirely sure if the controller is meant to control the actual state of the widget, given the other implementations of `MaterialStateController` in the framework https://github.com/flutter/flutter/pull/103167. If I try this out with 3 `InkWell` buttons, where 2 of the buttons update the `MaterialStateController` of the third button, adding states through `update` changes the internal `Set<MaterialState>` but does not actually update the visual state of the widget (the third button does not look focused/hovered when doing `update(.focused/.hovered, true)`).

Reply:
Talked with @HansMuller offline and it sounds like the primary purpose of `MaterialStatesController` is to control a widgets previously private `Set<MaterialState>`. This set of states is primarily used when resolving different `MaterialStateProperty` and does not control the widgets actual state mechanisms like focus/hovered etc. 

I think a user would expect that `MaterialStatesController` works the same on every widget that can be provided one, so i'm leaning towards including this functionality. What if we add documentation about `MaterialStatesController` not being able to control a widgets internal state mechanisms and that it will only change what `MaterialStateProperty`s resolve to?

Reply:
> I think a user would expect that MaterialStatesController works the same on every widget that can be provided one, so i'm leaning towards including this functionality.

It doesn't have to be a `MaterialStatesController` right? An `onChange` callback should suffice if all the user wants is to observe the currently hidden state.

> MaterialStatesController not being able to control a widgets internal state mechanisms and that it will only change what MaterialStatePropertys resolve to

Wouldn't this introduce a second source of truth for things like `.focused`? What does it mean when the `focusNode.hasFocus` and `materialStates` have different opinions about whether the text field should look like it's focused?

Reply:
> The purpose of this change is to let developers monitor material states changes, but changing the material states from outside of the `TextField` class shouldn't be allowed, since the source of the truth is `TextFieldState` I think?

FTR: adding a MaterialStateController property also enables one to extend the states that a component supports. Here's an example that creates a TextButton that supports the selected state:

```dart
import 'package:flutter/material.dart';

class SelectableButton extends StatefulWidget {
  const SelectableButton({
    super.key,
    required this.selected,
    this.style,
    required this.onPressed,
    required this.child,
  });

  final bool selected;
  final ButtonStyle? style;
  final VoidCallback? onPressed;
  final Widget child;

  bool get enabled => onPressed != null;

  @override
  State<SelectableButton> createState() => _SelectableButtonState();

}

class _SelectableButtonState extends State<SelectableButton> {
  late final MaterialStatesController statesController;

  @override
  void initState() {
    super.initState();
    statesController = MaterialStatesController({
      if (widget.selected) MaterialState.selected
    });
  }

  @override
  void didUpdateWidget(SelectableButton oldWidget) {
    super.didUpdateWidget(oldWidget);
    if (widget.selected != oldWidget.selected) {
      statesController.update(MaterialState.selected, widget.selected);
    }
  }

  @override
  Widget build(BuildContext context) {
    return TextButton(
      statesController: statesController,
      style: widget.style,
      onPressed: widget.onPressed,
      child: widget.child,
    );
  }
}

class Home extends StatefulWidget {
  const Home({ super.key });

  @override
  State<Home> createState() => _HomeState();
}

class _HomeState extends State<Home> {
  bool selected = false;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: SelectableButton(
          selected: selected,
          style: ButtonStyle(
            foregroundColor: MaterialStateProperty.resolveWith<Color?>(
              (Set<MaterialState> states) {
                if (states.contains(MaterialState.selected)) {
                  return Colors.white;
                }
                return null; // defer to the defaults
              },
            ),
            backgroundColor: MaterialStateProperty.resolveWith<Color?>(
              (Set<MaterialState> states) {
                if (states.contains(MaterialState.selected)) {
                  return Colors.indigo;
                }
                return null; // defer to the defaults
              },
            ),
          ),
          onPressed: () {
            setState(() { selected = !selected; });
          },
          child: const Text('toggle selected'),
        ),
      ),
    );
  }
}

void main() {
  runApp(const MaterialApp(home: Home()));
}
```

Reply:
So it sounds like each material state listens to 2 different input streams and the most recent value emitted by either stream dictates the value of that state?
But what does it mean if the user adds `.focused` to the states controller, and then the text field loses focus? If the owner of the controller wants to override the `.focused` state then `focusNode` shouldn't be able to also update the same state right?

Reply:
> So it sounds like each material state listens to 2 different input streams (the owner of the controller and the internal source of truth in TextFieldState) and the most recent value emitted by either stream dictates the value of that state?

That is correct. There's a subtle distinction here: the MaterialStatesController represents the visual state of the component, which typically tracks the intrinsic state of the component.

> But what does it mean if the user adds .focused to the states controller, and then the text field loses focus? If the owner of the controller wants to override the .focused state then focusNode shouldn't be able to also update the same state right?
If the user adds the focused state after the component loses focus, then the component's visuals will make it appear that the component has the focus. This isn't a new capability, you could always change a component's visual properties to obscure an underlying state change. Users typically don't do that, but you could.

Reply:
> If the user adds the focused MaterialState after the component loses focus, then the component's visuals will make it appear that the component has the focus.

And if the component regains focus and loses focus again? Should the visual state of the component update as the intrinsic state of the component changes? 
- If the answer is no then it sounds like we need an additional switch that the user can turn on to suppress the intrinsic state changes from affecting the visual state 
- otherwise the controller will be pretty hard to use on a `TextField`. For instance the app user can switch the focus to a different component, and to obscure the underlying state change I guess you'll have to listen to the state controller and change the material state back? That usually won't happen until the next frame. Average users may not notice that but that can make tests flaky.

Reply:
If you really wanted to prevent the MaterialStatesController from tracking the textfield's intrinsic focus (or whatever) state you could override the controller's update method. That said, this isn't a use case that has ever come up or seems worth designing for.

Reply:
Ah ok makes sense. I guess this should be in the documentation?

Reply:
Yes, it's not explained well (at all) and some examples are needed. If you create an issue, you can assign it to me.

Reply:
Updated the MaterialStatesController docs in https://github.com/flutter/flutter/pull/134592